ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/VersionedOffsetBuilder.scala

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import com.daml.ledger.offset.Offset
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, DataInputStream, DataOutputStream}
+import java.nio.ByteBuffer
+
+/** Helper functions for generating versioned 16 byte [[Offset]]s from integers.
+  * The created offset will look as follows:
+  * | version (8 bits) | highest index (56 bits) | middle index (32 bits) | lowest index (32 bits) |
+  * Leading zeros will be retained when generating the resulting offset bytes.
+  *
+  * Example usage:
+  *
+  *   - If you have one record per block then use [[of]] with the block ID as the second and last argument.
+  *   - If you may have multiple records per block then use [[of]] with the index within the block as the third argument.
+  *
+  * @see com.daml.ledger.offset.Offset
+  */
+class VersionedOffsetBuilder(version: Byte) {
+  import VersionedOffsetBuilder._
+
+  def onlyKeepHighestIndex(offset: Offset): Offset = {
+    val highest = highestIndex(offset)
+    of(highest)
+  }
+
+  def dropLowestIndex(offset: Offset): Offset = {
+    val (highest, middle, _) = split(offset)
+    of(highest, middle)
+  }
+
+  def setMiddleIndex(offset: Offset, middle: Int): Offset = {
+    val (highest, _, lowest) = split(offset)
+    of(highest, middle, lowest)
+  }
+
+  def setLowestIndex(offset: Offset, lowest: Int): Offset = {
+    val (highest, middle, _) = split(offset)
+    of(highest, middle, lowest)
+  }
+
+  def of(first: Long, second: Int = 0, third: Int = 0): Offset = {
+    if (first < 0 || first > MaxHighest)
+      throw new IllegalArgumentException(s"Highest: $first is out of range [0, $MaxHighest]")
```

Comment:
Shouldn't we also validate that `second` and `third` are positive?

Reply:
I think that's a good idea.

Reply:
Done